src/do_unmount/do_unmount.c

```diff
@@ -0,0 +1,382 @@
+ * except according to the terms contained in the LICENSE file.
+ */
+
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "helpers.h"
+#include "einfo.h"
+#include "queue.h"
+#include "rc.h"
+
+/* Provide a different failure message for each possible outcome */
+const char *failure_messages[] = {
+    "",                                     /* Placeholder for SUCCESS outcome */
+    "failed because we are using",          /* THIS */
+    "in use but fuser finds nothing",       /* UNKNOWN */
+    "in use but fuser command failed",      /* FUSER */
+    "in use but failed to kill process",    /* KILL */
+    "failed",                               /* ERROR */
+};
+
+/* Forward declaration to solve circular dependency */
+struct t_args_t;
+
+/* Contains parameters shared among all threads */
+typedef struct {
+    char* command;                /* Unmounting command to execute */
+    RC_STRINGLIST *shared;        /* List of shared paths in the system */
+    pthread_mutex_t shared_lock;  /* Mutex to lock shared mount operations */
+    struct t_args_t *args_array;  /* Array of all parameters of all threads */
+} global_args_t;
+
+/* Enumerate all the possible outcomes of the unmount procedure */
+typedef enum {
+    SUCCESS,  /* Unmounting was successful */
+    THIS,     /* The mount point is being used by this process */
+    UNKNOWN,  /* The mount point is being used but processes can't be found */
+    FUSER,    /* fuser command failed */
+    KILL,     /* The mount point is being used by other processes but the kill command failed */
+    ERROR,    /* Unknown error */
+} outcome_t;
+
+/* Contains all parameters passed to unmount_one() function */
+typedef struct t_args_t {
+    int index;                    /* Index of the path in the list */
+    RC_STRING *path;              /* Path to unmount */
+    pthread_mutex_t unmounting;   /* Mutex locked until the path is unmounted */
+    global_args_t *global_args;   /* Parameters shared among all threads */
+} thread_args_t;
+
+/* Pass arguments as vector of string to a command and open standard output as readable file */
+static FILE *
+popen_vec(const char *command, int argc, char **argv)
+{
+    FILE *fp;   /* File pointer to program output */
+    char *cmd;  /* Command with all arguments */
+    int length; /* Length of the command with all arguments */
+    int i;      /* Iterator */
+
+    /* Calculate the length of the command */
+    length = strlen(command) + 1; /* +1 for the null terminator */
+    for (i = 0; i < argc; i++)
+        length += strlen(argv[i]) + 3; /* +3 for the quotes and space */
+
+    /* Allocate memory for the command */
+    cmd = xmalloc(length * sizeof(char));
+    sprintf(cmd, "%s", command);
+
+    /* Craft the command with all passed arguments */
+    for (i = 0; i < argc; i++) {
+        strcat(cmd, " \"");
+        strcat(cmd, argv[i]);
+        strcat(cmd, "\"");
```

Comment:
This should *really* be doing fork+exec (or `posix_spawn()` if you're familiar with that). This type of simplistic escapes are not robust, will fail if the argument contains special character or worse execute random stuff in case of `$(...)` or backticks.